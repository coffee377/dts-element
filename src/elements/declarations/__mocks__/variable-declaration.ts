import {Declaration} from '../../declaration';

export class VariableDeclaration extends Declaration<any, any> {

  public get default_parameters(): any {
    return super.default_declaration_parameters;
  }

  public _emit_raw(): string {
    return `[VariableDeclaration ${this.parameters.name}]`;
  }

}
