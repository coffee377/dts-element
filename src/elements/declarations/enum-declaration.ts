import {emit_declare} from '../../helpers/emit-declare';
import {emit_enum_members} from '../../helpers/emit-enum-members';
import {Stack} from '../../stack';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {EnumMember} from '../members/enum-member';

export interface IEnumDeclarationRequiredParameters {
  name: string;
}

export interface IEnumDeclarationOptionalParameters {
  members: EnumMember[];
}

export class EnumDeclaration
    extends Declaration<IEnumDeclarationRequiredParameters, IEnumDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & IEnumDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      members: [],
    });
  }

  public _emit_raw(stack: Stack): string {
    const {name, members} = this.parameters;
    const declare = emit_declare(stack);
    return `${declare}enum ${name} ${emit_enum_members(members, stack)}`;
  }

}
