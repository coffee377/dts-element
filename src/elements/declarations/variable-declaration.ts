import {Container} from '../../collections';
import {any_type} from '../../constants';
import {emit_declaration_name} from '../../helpers/emit-declaration-name';
import {emit_declare} from '../../helpers/emit-declare';
import {emit_optional} from '../../helpers/emit-optional';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {Type} from '../type';

export type VariableKind = 'var' | 'let' | 'const';

// tslint:disable-next-line no-empty-interface
export interface IVariableDeclarationRequiredParameters {}

export interface IVariableDeclarationOptionalParameters {
  kind: VariableKind;
  type: Type;
  optional: boolean;
}

export class VariableDeclaration
    extends Declaration<IVariableDeclarationRequiredParameters, IVariableDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & IVariableDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      kind: 'var' as VariableKind,
      type: any_type,
      optional: false,
    });
  }

  public _emit_raw(container: Container): string {
    const {kind, type} = this.parameters;
    const name = emit_declaration_name(this.parameters.name, container);
    const optional = emit_optional(this.parameters.optional);
    const declare = emit_declare(container);
    return `${declare}${kind} ${name}${optional}: ${type._emit(this)};`;
  }

}
