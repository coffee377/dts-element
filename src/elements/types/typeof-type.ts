import {Stack} from '../../stack';
import {Declaration} from '../declaration';
import {Type} from '../type';

export interface ITypeofTypeRequiredParameters {
  owned: Declaration;
}

// tslint:disable-next-line no-empty-interface
export interface ITypeofTypeOptionalParameters {}

export class TypeofType extends Type<ITypeofTypeRequiredParameters, ITypeofTypeOptionalParameters> {

  public get default_parameters(): ITypeofTypeOptionalParameters {
    return {};
  }

  public _emit(_stack: Stack): string {
    const {owned} = this.parameters;
    return `typeof ${owned.parameters.name}`;
  }

}
