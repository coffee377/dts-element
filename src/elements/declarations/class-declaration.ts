import {Container} from '../../collections';
import {emit_abstract} from '../../helpers/emit-abstract';
import {emit_extends} from '../../helpers/emit-extends';
import {emit_generics} from '../../helpers/emit-generics';
import {emit_members} from '../../helpers/emit-members';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {ClassMember} from '../members/class-member';
import {ClassType} from '../types/class-type';
import {GenericType} from '../types/generic-type';

export interface IClassDeclarationRequiredParameters {
  name: string;
}

export interface IClassDeclarationOptionalParameters {
  generics: GenericType[];
  members: ClassMember[];
  extends: ClassType | null;
  abstract: boolean;
}

export class ClassDeclaration
    extends Declaration<IClassDeclarationRequiredParameters, IClassDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & IClassDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      generics: [],
      members: [],
      extends: null,
      abstract: false,
    });
  }

  public _emit_raw(container: Container): string {
    const {name, members, generics} = this.parameters;
    const generic = emit_generics(generics, this);
    const abstract = emit_abstract(this.parameters.abstract, this);
    const an_extends = emit_extends(this.parameters.extends, this);
    return `${abstract}class ${name}${generic}${an_extends} ${emit_members(members, this)}`;
  }

}
