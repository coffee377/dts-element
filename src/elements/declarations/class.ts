import {Container} from '../../collections';
import {emit_abstract} from '../../helpers/emit-abstract';
import {emit_declaration_name} from '../../helpers/emit-declaration-name';
import {emit_extends} from '../../helpers/emit-extends';
import {emit_generics} from '../../helpers/emit-generics';
import {emit_object} from '../../helpers/emit-object';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {ClassMember} from '../members/class';
import {ClassType} from '../types/class';
import {GenericType} from '../types/generic';

export interface IClassDeclarationRequiredParameters {
  name: string;
}

export interface IClassDeclarationOptionalParameters {
  generics: GenericType[];
  children: ClassMember[];
  extends: ClassType | null;
  abstract: boolean;
}

export class ClassDeclaration
    extends Declaration<IClassDeclarationRequiredParameters, IClassDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & IClassDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      generics: [],
      children: [],
      extends: null,
      abstract: false,
    });
  }

  public _emit_raw(container: Container): string {
    const {children, generics} = this.parameters;
    const name = emit_declaration_name(this.parameters.name, container);
    const generic = emit_generics(generics, this);
    const abstract = emit_abstract(this.parameters.abstract, this);
    const an_extends = emit_extends(this.parameters.extends, this);
    return `${abstract}class ${name}${generic}${an_extends} ${emit_object(children, this)}`;
  }

}
