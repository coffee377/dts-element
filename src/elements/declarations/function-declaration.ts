import {any_type} from '../../constants';
import {emit_declaration_name} from '../../helpers/emit-declaration-name';
import {emit_declare} from '../../helpers/emit-declare';
import {emit_export} from '../../helpers/emit-export';
import {emit_function} from '../../helpers/emit-function';
import {emit_generics} from '../../helpers/emit-generics';
import {emit_parameters} from '../../helpers/emit-parameters';
import {Stack} from '../../stack';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {Parameter} from '../parameter';
import {Type} from '../type';
import {TypeAssertion} from '../type-assertion';
import {FunctionType} from '../types/function-type';
import {GenericType} from '../types/generic-type';

// tslint:disable-next-line no-empty-interface
export interface IFunctionDeclarationRequiredParameters {}

export interface IFunctionDeclarationOptionalParameters {
  type: FunctionType;
  export: boolean;
}

export class FunctionDeclaration
    extends Declaration<IFunctionDeclarationRequiredParameters, IFunctionDeclarationOptionalParameters> {

  private _instance_of_function_declaration: true;

  public get default_parameters(): IDeclarationOptionalParameters & IFunctionDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      type: new FunctionType({return: any_type}),
      export: false,
    });
  }

  public _emit_raw(stack: Stack): string {
    const {type} = this.parameters;
    const name = emit_declaration_name(this.parameters.name, stack);
    const parameters = emit_parameters(type.parameters.parameters, stack);
    const generics = emit_generics(type.parameters.generics, stack, true);
    const return_type = type.parameters.return.emit(stack);
    const declare = emit_declare(stack);
    const func = emit_function(stack);
    const an_export = emit_export(this.parameters.export, stack);
    return `${an_export}${declare}${func}${name}${generics}(${parameters}): ${return_type};`;
  }

}
