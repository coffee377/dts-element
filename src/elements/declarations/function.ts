import {any_type} from '../../constants';
import {AnyElement} from '../../element';
import {emit_declare} from '../../helpers/emit-declare';
import {emit_function} from '../../helpers/emit-function';
import {emit_generics} from '../../helpers/emit-generics';
import {emit_parameters} from '../../helpers/emit-parameters';
import {Declaration, IDeclarationOptionalParameters} from '../declaration';
import {Parameter} from '../parameter';
import {AnyType} from '../type';
import {GenericType} from '../types/generic';

// tslint:disable-next-line no-empty-interface
export interface IFunctionDeclarationRequiredParameters {}

export interface IFunctionDeclarationOptionalParameters {
  parameters: Parameter[];
  generics: GenericType[];
  return: AnyType;
}

export class FunctionDeclaration
    extends Declaration<IFunctionDeclarationRequiredParameters, IFunctionDeclarationOptionalParameters> {

  public get default_parameters(): IDeclarationOptionalParameters & IFunctionDeclarationOptionalParameters {
    return Object.assign({}, super.default_declaration_parameters, {
      parameters: [],
      generics: [],
      return: any_type,
    });
  }

  public _emit(container: AnyElement | null): string {
    const {name} = this.parameters;
    const parameters = emit_parameters(this.parameters.parameters, this);
    const generics = emit_generics(this.parameters.generics, this);
    const return_type = this.parameters.return._emit(this);
    const declare = emit_declare(container);
    const func = emit_function(container);
    return `${declare}${func}${name}${generics}(${parameters}): ${return_type};`;
  }

}
