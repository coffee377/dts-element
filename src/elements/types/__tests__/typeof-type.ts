jest.unmock('../typeof-type');

import {VariableDeclaration} from '../../declarations/variable-declaration';
import {TypeofType} from '../typeof-type';

describe('#emit()', () => {
  it('should return correctly', () => {
    expect(new TypeofType({
      owned: new VariableDeclaration({name: 'test'}),
    }).emit()).toMatchSnapshot();
  });
});
