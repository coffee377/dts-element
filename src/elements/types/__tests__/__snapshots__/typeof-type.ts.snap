// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#emit() should return correctly 1`] = `"typeof test"`;
